### PR TITLE
[Calling][Feature] Custom buttons

### PR DESCRIPTION
**Pull request by pavelprystinka**
## Purpose

* Custom buttons feature

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

## Pull Request Type
What kind of change does this Pull Request introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Pull Request Checklist

- [X] Public API changes
- [ ] Verified for cross-platform
- [ ] Synced with iOS, Web team
- [ ] Internal review completed
- [ ] UI Changes
  - [ ] Screen captures included
  - [ ] Tested for Light/Dark mode
  - [ ] Tested for screen rotation
  - [ ] Tested on Tablet and small screen device (5")
  - [ ] Include localization changes
- [ ] Tests
  - [ ] Memory leak analysis performed
  - [ ] Tested on API 21, 26 and latest 
  - [ ] Tested for background mode
  - [ ] Unit Tests Included
  - [ ] UI Automated Tests Included

## How to Test

* Open...
* Click on...

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: azure-communication-ui/calling/publish.gradle [hunk 1]
```diff
@@ -59,56 +59,5 @@ publishing {
                 from components.release
             }
         }
-
```
Comment (ahammer) [inline azure-communication-ui/calling/publish.gradle:62]: Why is this removed as part of this PR?

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/models/CallCompositeCustomButtonOptions.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * Call composite custom button options.
+ */
+public final class CallCompositeCustomButtonOptions {
+
+    private final int drawableId;
+    private final String title;
+    private final CallCompositeEventHandler<CallCompositeCustomButtonClickEvent> onClickHandler;
+    private Boolean isEnabled = true;
```
Comment (ahammer) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/models/CallCompositeCustomButtonOptions.java:16]: Why mix of immutable constructor, and mutable isEnabled?

If isEnabled can be changed later, shouldn't the rest also be updatable?

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/ControlBarViewModel.kt [hunk 3]
```diff
@@ -216,4 +269,16 @@ internal class ControlBarViewModel(
...
+
+    private fun callOnClickHandler(
+        context: Context,
+        buttonOptions: CallCompositeButtonOptions?,
+    ) {
+        try {
+            buttonOptions?.onClickHandler?.handle(
+                createButtonClickEvent(context, buttonOptions)
+            )
+        } catch (_: Exception) {
```
Comment (ahammer) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/ControlBarViewModel.kt:281]: Should Log.e so developer can see error we are squelching in Logcat.

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/captions/CaptionsListView.kt [hunk 4]
```diff
@@ -123,54 +124,69 @@ internal class CaptionsContainerView(
...
+                    BottomCellItemType.BottomMenuAction,
+                    showRightArrow = true,
+                    subtitle = LocaleHelper.getLocaleDisplayName(activeSpokenLanguage),
+                    onClickAction = {
+                        try {
+                            viewModel.spokenLanguageButtonOptions?.onClickHandler?.handle(
+                                createButtonClickEvent(context, viewModel.spokenLanguageButtonOptions)
+                            )
+                        } catch (_: Exception) {
+                        }
```
Comment (ahammer) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/captions/CaptionsListView.kt:182]: same as above, Log exceptions we hide so dev can see them in logcat.

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/captions/CaptionsListView.kt [hunk 5]
```diff
@@ -189,9 +205,15 @@ internal class CaptionsContainerView(
                     showRightArrow = true,
                     subtitle = LocaleHelper.getLocaleDisplayName(activeCaptionLanguage),
                     onClickAction = {
+                        try {
+                            viewModel.captionsLanguageButtonOptions?.onClickHandler?.handle(
+                                createButtonClickEvent(context, viewModel.captionsLanguageButtonOptions)
+                            )
+                        } catch (_: Exception) {
+                        }
```
Comment (ahammer) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/captions/CaptionsListView.kt:213]: ^

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/more/MoreCallOptionsListViewModel.kt [hunk 6]
```diff
@@ -27,14 +39,69 @@ internal class MoreCallOptionsListViewModel(
...
+                        onClickListener = { context ->
+                            try {
+                                customButton.onClickHandler?.handle(
+                                    createCustomButtonClickEvent(
+                                        context,
+                                        customButton
+                                    )
+                                )
+                            } catch (_: Exception) {
+                            }
```
Comment (ahammer) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/more/MoreCallOptionsListViewModel.kt:100]: ^

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/more/MoreCallOptionsListViewModel.kt [hunk 7]
```diff
@@ -54,25 +121,31 @@ internal class MoreCallOptionsListViewModel(
...
-            ),
+    private fun callOnClickHandler(
+        context: Context,
+        buttonOptions: CallCompositeButtonOptions?,
+    ) {
+        try {
+            buttonOptions?.onClickHandler?.handle(
+                createButtonClickEvent(context, buttonOptions)
+            )
+        } catch (_: Exception) {
```
Comment (ahammer) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/more/MoreCallOptionsListViewModel.kt:138]: ^

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarView.kt [hunk 8]
```diff
@@ -201,17 +183,17 @@ internal class SetupControlBarView : LinearLayout {
 
     private fun toggleAudio() {
         if (micButton.isON) {
-            viewModel.turnMicOff()
+            viewModel.turnMicOff(context)
```
Comment (ahammer) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarView.kt:186]: This business logic should be pushed up to the VM I think.

I.e.
instead of 
        micButton.setOnClickListener {
            toggleAudio()
        }
       
it should be
micButton.setOnClickListener(viewModel::onMicClicked)

And the logic about what to do from there (i.e. mic on/off etc, should live in the VM).

i.e.

onMicClicked(context) {
  dispatchCorrectActionForMicNow()
  notifyMicClickedListener()
}

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarViewModel.kt [hunk 9]
```diff
@@ -87,41 +102,50 @@ internal class SetupControlBarViewModel(
...
-    val deviceIsEnabled: StateFlow<Boolean> get() = deviceIsEnabledStateFlow
+    val audioDeviceButtonEnabled: StateFlow<Boolean> get() = audioDeviceButtonIsEnabledStateFlow
     val isVisibleState: StateFlow<Boolean> get() = visibleStateFlow
     val cameraState: StateFlow<CameraOperationalStatus> get() = cameraStateFlow
 
     val audioOperationalStatusStat: StateFlow<AudioOperationalStatus> get() = audioOperationalStatusStateFlow
     val audioDeviceSelectionStatusState: StateFlow<AudioState> get() = audioDeviceSelectionStatusStateFlow
 
-    fun turnCameraOn() {
+    fun turnCameraOn(context: Context) {
```
Comment (ahammer) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarViewModel.kt:116]: as per above, this would look better if they were direct bindings for the button, and had the logic that is in the View be in the ViewModel instead for easier testing.

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarViewModel.kt [hunk 10]
```diff
@@ -153,12 +179,32 @@ internal class SetupControlBarViewModel(
...
         return !shouldControlsBeDisabled(callingState) &&
             audioStateOperation != AudioOperationalStatus.PENDING &&
-            setupScreenOptions?.isMicrophoneButtonEnabled != false
+            setupScreenOptions?.microphoneButton?.isEnabled != false
+    }
+
+    private fun shouldAudioDeviceButtonBeEnabled(
+        callingState: CallingState,
+        setupScreenOptions: CallCompositeSetupScreenOptions?
+    ): Boolean {
```
Comment (ahammer) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarViewModel.kt:188]: nit: Kotlin shorthand

fun something() -> true/false
Type inferrence, single line collapse to statement.

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarViewModel.kt [hunk 11]
```diff
@@ -153,12 +179,32 @@ internal class SetupControlBarViewModel(
...
+
+    private fun callOnClickHandler(
+        context: Context,
+        buttonOptions: CallCompositeButtonOptions?,
+    ) {
+        try {
+            buttonOptions?.onClickHandler?.handle(
+                createButtonClickEvent(context, buttonOptions)
+            )
+        } catch (_: Exception) {
```
Comment (ahammer) [inline azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarViewModel.kt:207]: don't hide exceptions, log.e

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/models/CallCompositeCustomButtonOptions.java (hunk 2 above)
Comment (pavelprystinka) [reply azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/models/CallCompositeCustomButtonOptions.java:16]: In the constructor we take a must to provide parameters: icon, tittle and onclick. 
isEnabled is optional so we make it assignable via setter.

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/ControlBarViewModel.kt (hunk 3 above)
Comment (pavelprystinka) [reply azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/ControlBarViewModel.kt]: Added

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/captions/CaptionsListView.kt (hunk 4 above)
Comment (pavelprystinka) [reply azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/captions/CaptionsListView.kt]: Added

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/captions/CaptionsListView.kt (hunk 5 above)
Comment (pavelprystinka) [reply azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/captions/CaptionsListView.kt]: Added

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/more/MoreCallOptionsListViewModel.kt (hunk 6 above)
Comment (pavelprystinka) [reply azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/more/MoreCallOptionsListViewModel.kt:103]: Added

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/more/MoreCallOptionsListViewModel.kt (hunk 7 above)
Comment (pavelprystinka) [reply azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/calling/controlbar/more/MoreCallOptionsListViewModel.kt]: Added

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarViewModel.kt (hunk 11 above)
Comment (pavelprystinka) [reply azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarViewModel.kt]: Added

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarView.kt (hunk 8 above)
Comment (pavelprystinka) [reply azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarView.kt:186]: I think this is the actual button state logic that should stay in the view.
ViewModel should not keep that state, but putting this state on Redux is a bit too much.
It was this way already.

File: azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarViewModel.kt (hunk 10 above)
Comment (pavelprystinka) [reply azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/presentation/fragment/setup/components/SetupControlBarViewModel.kt:190]: I do use `fun something() -> true/false` in simple cases. I find it hard to read if return is more complex.
For instance, `fun something() -> isConditions1 && isCondition2` then I use shorthand. 
But in this case I find it more cluttered and hurting readability.

**Review by ahammer — APPROVED** [review]

**Review by ShaunaSong — APPROVED** [review]
